src/core/server/elasticsearch/cluster_client.ts

```diff
@@ -0,0 +1,182 @@
+    if (!options.wrap401Errors || err.statusCode !== 401) {
+      throw err;
+    }
+
+    const boomError = Boom.boomify(err, { statusCode: err.statusCode });
+    const wwwAuthHeader: string = get(err, 'body.error.header[WWW-Authenticate]');
+    boomError.output.headers['WWW-Authenticate'] =
+      wwwAuthHeader || 'Basic realm="Authorization Required"';
+
+    throw boomError;
+  }
+}
+
+/**
+ * Represents an Elasticsearch cluster API client and allows to call API on behalf
+ * of the internal Kibana user and the actual user that is derived from the request
+ * headers (via `asScoped(...)`).
+ */
+export class ClusterClient {
+  /**
+   * Raw Elasticsearch JS client that acts on behalf of the Kibana internal user.
+   */
+  private readonly client: Client;
+
+  /**
+   * Optional raw Elasticsearch JS client that is shared between all the scoped clients created
+   * from this cluster client. Every API call is attributed by the wh
+   */
+  private scopedClient?: Client;
+
+  /**
+   * Indicates whether this cluster client (and all internal raw Elasticsearch JS clients) has been closed.
+   */
+  private isClosed = false;
+
+  constructor(private readonly config: ElasticsearchClientConfig, private readonly log: Logger) {
+    this.client = new Client(parseElasticsearchClientConfig(config, log));
+  }
+
+  /**
+   * Calls specified {@param endpoint} with provided {@param clientParams} on behalf of the
+   * Kibana internal user.
+   * @param endpoint String descriptor of the endpoint e.g. `cluster.getSettings` or `ping`.
+   * @param clientParams A dictionary of parameters that will be passed directly to the Elasticsearch JS client.
+   * @param options Options that affect the way we call the API and process the result.
+   */
+  public async callAsInternalUser(
+    endpoint: string,
+    clientParams: Record<string, unknown> = {},
+    options?: CallAPIOptions
+  ) {
+    this.assertIsNotClosed();
+
+    return await callAPI(this.client, endpoint, clientParams, options);
+  }
+
+  /**
+   * Closes the cluster client. After that client cannot be used and one should
+   * create a new client instance to be able to interact with Elasticsearch API.
+   */
+  public close() {
+    this.client.close();
+
+    if (this.scopedClient !== undefined) {
+      this.scopedClient.close();
+    }
+
+    this.isClosed = true;
+  }
+
+  /**
+   * Creates an instance of `ScopedClusterClient` based on the configuration the
+   * current cluster client that exposes additional `callAsCurrentUser` method
+   * scoped to the provided {@param req}. Consumers shouldn't worry about closing
+   * scoped client instances, these will be automatically closed as soon as the
+   * original cluster client isn't needed anymore and closed.
+   * @param req Request the `ScopedClusterClient` instance will be scoped to.
+   */
+  public asScoped(req: { headers?: Headers } = {}) {
+    // It'd have been quite expensive to create and configure client for every incoming
```

Comment:
__note:__ I tried to explain in comment why scoped cluster client was implemented in this way. I spent some time learning how ES plugin works currently and is being used and couldn't find a better way to achieve that, but happy to hear your suggestions.